dali/python/nvidia/dali/external_source.py

```diff
@@ -296,13 +330,27 @@ class ExternalSource():
     of separate Tensors, there will be an additional copy made internally, consuming both memory
     and bandwidth.
 
+    Cannot be explicitly set to ``False`` when used with ``parallel = True``
+
 `batch` : bool, optional
     If set to ``True`` or ``None``, the ``source`` is expected to produce an entire batch at once.
     If set to ``False``, the ``source`` is called per-sample.
+
+`parallel` : bool, optional, default = False
+    If set to ``True`` the corresponding pipeline will run pool of Python workers to run the
+    callback in parallel. You can specify the number of workers by passing ``py_workers_num``
+    into pipeline's constructor.
+
+    When parallel=True ``source`` must return NumPy ndarray (or tuple of arrays matching num_outputs).
```

Comment:
What happens when it returns cupy or TensorList, PyTorch tensor and etc?

Reply:
We can probably make it anything compatible with `__array_interface__`, but I don't think we want to try CuPy now. Let's focus on CPU variant first.

Reply:
TypeError is raised. It happens in shared_batch.py/_add_sample_to_batch when encountering something that is neither a primitive, numpy array nor tuple/list of them.
I believe it should not be particularly complicated to extend supported types (either through array_interface approach or just trying to cast everything (on cpu) to numpy arrays), but my suggestion would be, due to little time there is and the scope of the PR to make it a follow up, what do you think?